### ISSUE TITLE
WebXR Layers

### ISSUE THREAD

**Issue by cabanier** [w3ctag/design-reviews#528]
Saluton TAG!

I'm requesting a TAG review of the [WebXR Layers specification](https://immersive-web.github.io/layers/).

WebXR Layers offers a more efficient way of drawing immersive content than WebXR. In addition to support for native textures and texture arrays, it also provides support for different layer types that are managed by the system compositor (as opposed to javascript). This spec is an optional module that extends the feature set of [WebXR](https://immersive-web.github.io/webxr).

  - Specification URL: https://immersive-web.github.io/layers/
  - Explainer: https://github.com/immersive-web/layers/blob/main/explainer.md
  - Security and Privacy self-review: https://github.com/immersive-web/layers/blob/main/security-considerations.md
  - GitHub repo (if you prefer feedback filed there): https://github.com/immersive-web/layers/
  - Primary contacts:
      - [Rik Cabanier]([cabanier]), [Facebook](https://www.facebook.com/)
  - Organization/project driving the design: [Oculus](https://www.oculus.com/) ([Facebook](https://www.facebook.com/))
  - External status/issue trackers for this feature: 
    - [Chrome Status](https://chromestatus.com/admin/features/edit/6287363875536896)
    - [Chrome Tracking bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1096923)

Further details:

  - [x] I have reviewed the TAG's [API Design Principles](https://w3ctag.github.io/design-principles/)
  - The group where the incubation/design work on this is being done (or is intended to be done in the future): [W3C Immersive Web CG](https://www.w3.org/community/immersive-web/)
  - The group where standardization of this work is intended to be done: [W3C Immersive Web WG](https://github.com/immersive-web)
  - Major unresolved issues with or opposition to this design: none
  - This work is being funded by: [Facebook](https://www.facebook.com/)

We'd prefer the TAG provide feedback as:
  🐛 open issues in our GitHub repo for **each point of feedback**

**Comment by torgo** [comment]
Hi @cabanier - We are just in the process of reviewing in our virtual f2f.  One quick suggestion: I feel like the explainer would be a bit more explainy if you took the firs paragraph from the "use cases" section and moved it to the very top of the document – that way it could provide important context for what I'm reading right under it.

**Comment by torgo** [comment]
Hi  @cabanier. @atanassov and I are just reviewing now in a special break-out. I noticed that you didn't answer the security & privacy questionnaire, opting instead to point us to the spec, but in the spec there is only a very short privacy & consideration section. Can you please try to answer the actual security & privacy considerations questions (in a separate doc - doesn't need to be fancy - a text or markdown doc is fine)?  I think there may be more factors you need to consider then what you have listed in the considerations section.

**Comment by cabanier** [comment]
Thanks Dan!
I will do so today.

**Comment by atanassov** [comment]
@cabanier one more editorial feedback about the explainer. The two introduction sections offer little motivation to users who aren't deeply familiar with WebXR, rendering or composition. It will go a long way if you can add a bit more details to that end.

Towards the end of the explainer there is usually a section that describes other approaches that were considered and why this is one the best. Given there are so many capabilities in this spec I can see why that could be difficult and hope you can offer at least some view into your decision making.

**Comment by atanassov** [comment]
@cabanier couple of more questions as we read the spec. 

It seems that exposing `nativeProjectionScaleFactor` will enable authors using the feature to obtain the scale factor for a given context (assuming post transform?). If this is the case, my assumption is that such capability will be very sought after many other places in the platform (users of CSSOM have been asking for something like this forever). Could this be either omitted or made available outside the WebXR context?

Have you thought through some of the threat vectors that could abuse such capability? How quickly and easily could a malicious actor create a WebXR context in a different content and obtain the composition layers?

One final meta question. Have you considered, or at least do you think it is feasible to break down this review into few logical pieces that build on top of each other? Layer Types, Spaces, Creation, Rendering etc. That model of reviews will be much more expedient to all of us.

**Comment by cabanier** [comment]
> It seems that exposing `nativeProjectionScaleFactor` will enable authors using the feature to obtain the scale factor for a given context (assuming post transform?). If this is the case, my assumption is that such capability will be very sought after many other places in the platform (users of CSSOM have been asking for something like this forever). Could this be either omitted or made available outside the WebXR context?

`nativeProjectionScaleFactor` is the value of [`scaleFactor`](https://immersive-web.github.io/layers/#dom-xrprojectionlayerinit-scalefactor) that was passed in during the construction of the projection layer.
It's basically the same as the [`framebufferscalefactor`](https://www.w3.org/TR/webxr/#dom-xrwebgllayerinit-framebufferscalefactor) value in the WebXR spec
It is NOT to the "scale factor for a given context (assuming post transform?)". WebXR Layers can only be used in immersive sessions so it has no relation with what is rendered on the page.

> Have you thought through some of the threat vectors that could abuse such capability? 

There's no new capability with the scale factor. The system will just allocate a buffer where the native buffer size is scaled by this number.

> How quickly and easily could a malicious actor create a WebXR context in a different content and obtain the composition layers?

It wouldn't be possible. Immersive WebXR sessions can only be created after a user action and consenting to go immersive. This is not something new that the WebXR Layers spec introduces; this is [already covered](https://www.w3.org/TR/webxr/#dom-xrsystem-requestsession) in the WebXR spec.

> One final meta question. Have you considered, or at least do you think it is feasible to break down this review into few logical pieces that build on top of each other? Layer Types, Spaces, Creation, Rendering etc. That model of reviews will be much more expedient to all of us.

As I mentioned earlier, this spec heavily depends on the WebXR specification so in order to review it, one has to be familiar with that one as well. 
I am planning on adding examples and non-normative explainers to clarify the intent. Right now it's mostly a list of interfaces and flow diagrams which makes it hard to follow.

**Comment by cabanier** [comment]
>  Can you please try to answer the [actual security & privacy considerations](https://www.w3.org/TR/security-privacy-questionnaire/) questions (in a separate doc - doesn't need to be fancy - a text or markdown doc is fine)? 

I added a markdown document [here](https://github.com/immersive-web/layers/blob/master/security-considerations.md).

> I think there may be more factors you need to consider then what you have listed in the considerations section.

Apart from timing of the composition, I'm unsure if there are any other considerations. The spec at its most basic level just exposes new textures to render content to.
I'd be happy to have a conversation or to add more to the specification to clarify.

**Comment by cabanier** [comment]
@atanassov @torgo Do I need to do anything else to have this picked up again by the TAG?

**Comment by torgo** [comment]
@cabanier we're just reviewing in today's TAG call. Thanks for filling out the security & privacy questionnaire - much appreciated. We have no red flags per se, but this is such a big piece of work that we're worried we're missing something. We could do with some more detail in the explainer - specifically "goals and non-goals, other potential solutions considered," also it's not clear to us which parts of the examples are newly added capabilities vs new functionality. Would it be possible for you to join one of our calls in the new year to go through it?

**Comment by cabanier** [comment]
Certainly! I will work on cleaning up the explainer.

**Comment by cabanier** [comment]
@torgo FYI I added "goals and non-goals, other potential solutions considered" to the spec.

**Comment by torgo** [comment]
Thanks for this Rik - we are generally positive - we're going to spend a little bit more time on the security & privacy review and we will try to close this one off very early in the new year.

**Comment by atanassov** [comment]
@cabanier, we had another look at the security & privacy document, thanks for completing. One question and concern that came up (security more than privacy) is around the lifetime management of layers. Can you elaborate on the coupling between the layer objects and underlying GPU textures or other lower level primitives? How is this proposal ensuring that lower primitives are protected from malicious use? 

Ideally this can go directly into the the security & privacy document or the explainer.

**Comment by cabanier** [comment]
@atanassov can you elaborate your concern a bit?The textures are defined to only be valid during a Raf call and after a call to  `getSubImage`/`getViewSubImage`. If you use them at any other time, they are considered invalid.

**Comment by cabanier** [comment]
ping @atanassov

**Comment by atanassov** [comment]
@cabanier sorry about the delay. @torgo and myself had this scheduled for completion during our current "Kronos" VF2F.

The textures explanation above is what I was missing. If you don't mind adding it to the security section of the spec and/or the privacy and security document.

With that added we are happy to close the review as satisfied.

**Comment by cabanier** [comment]
Thanks @atanassov!
I updated the explainer with that information and added some more details. Let me know if you still have questions or concerns.

**Comment by atanassov** [comment]
Thank you @cabanier! Took me a bit to see the change in your explainer only to realize the link here was to 'maser' not 'main' branch :) 
Anyway, everything looks good and we are excited to see progress being made at  W3C Immersive Web WG. Thank you for opening this review and working with us, good luck.
